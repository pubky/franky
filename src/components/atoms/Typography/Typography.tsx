import { ReactNode } from 'react';
import { cn } from '@/libs';

interface TypographyProps {
  as?:
    | 'h1'
    | 'h2'
    | 'h3'
    | 'h4'
    | 'h5'
    | 'h6'
    | 'p'
    | 'span'
    | 'i'
    | 'b'
    | 'strong'
    | 'em'
    | 'small'
    | 'sub'
    | 'sup'
    | 'code'
    | 'pre'
    | 'blockquote';
  children: ReactNode;
  className?: string;
  size?: 'sm' | 'md' | 'lg' | 'xl';
  'data-testid'?: string;
}

export function Typography({
  as: Tag = 'p',
  children,
  className,
  size = 'md',
  'data-testid': dataTestId,
  ...props
}: TypographyProps & React.HTMLAttributes<HTMLElement>) {
  const sizeClasses = {
    sm: 'text-lg font-semibold',
    md: 'text-xl font-semibold',
    lg: 'text-2xl font-bold',
    xl: 'text-4xl font-bold',
  };

  return (
    <Tag
      data-testid={dataTestId || 'typography'}
      className={cn(sizeClasses[size], 'text-foreground', className)}
      {...props}
    >
      {children}
    </Tag>
  );
}
