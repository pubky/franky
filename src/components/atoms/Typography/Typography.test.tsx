import { describe, it, expect } from 'vitest';
import { render, screen } from '@testing-library/react';
import { Typography } from './Typography';

describe('Typography', () => {
  it('renders with default props', () => {
    render(<Typography>Default text</Typography>);
    const typography = screen.getByText('Default text');
    expect(typography).toBeInTheDocument();
  });
});

describe('Typography - Snapshots', () => {
  it('matches snapshot with default props', () => {
    const { container } = render(<Typography>Default text</Typography>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for h1 element', () => {
    const { container } = render(<Typography as="h1">H1 heading</Typography>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for h2 element', () => {
    const { container } = render(<Typography as="h2">H2 heading</Typography>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for span element', () => {
    const { container } = render(<Typography as="span">Span text</Typography>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for strong element', () => {
    const { container } = render(<Typography as="strong">Strong text</Typography>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for em element', () => {
    const { container } = render(<Typography as="em">Emphasized text</Typography>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for small size', () => {
    const { container } = render(<Typography size="sm">Small text</Typography>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for medium size', () => {
    const { container } = render(<Typography size="md">Medium text</Typography>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for large size', () => {
    const { container } = render(<Typography size="lg">Large text</Typography>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot for extra large size', () => {
    const { container } = render(<Typography size="xl">Extra large text</Typography>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with custom className', () => {
    const { container } = render(<Typography className="custom-typography">Custom typography</Typography>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with combined props', () => {
    const { container } = render(
      <Typography as="h2" size="lg">
        Large heading
      </Typography>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with complex children', () => {
    const { container } = render(
      <Typography>
        <span>Complex</span> <em>typography</em> content
      </Typography>,
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with id prop', () => {
    const { container } = render(<Typography id="typography-id">Typography with ID</Typography>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('matches snapshot with data-testid prop', () => {
    const { container } = render(<Typography data-testid="custom-typography">Typography with test ID</Typography>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
