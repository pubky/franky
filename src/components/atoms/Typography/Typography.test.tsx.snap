// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Typography - Snapshots > matches snapshot for em element 1`] = `
<em
  class="text-xl font-semibold text-foreground"
  data-testid="typography"
>
  Emphasized text
</em>
`;

exports[`Typography - Snapshots > matches snapshot for extra large size 1`] = `
<p
  class="text-4xl font-bold text-foreground"
  data-testid="typography"
>
  Extra large text
</p>
`;

exports[`Typography - Snapshots > matches snapshot for h1 element 1`] = `
<h1
  class="text-xl font-semibold text-foreground"
  data-testid="typography"
>
  H1 heading
</h1>
`;

exports[`Typography - Snapshots > matches snapshot for h2 element 1`] = `
<h2
  class="text-xl font-semibold text-foreground"
  data-testid="typography"
>
  H2 heading
</h2>
`;

exports[`Typography - Snapshots > matches snapshot for large size 1`] = `
<p
  class="text-2xl font-bold text-foreground"
  data-testid="typography"
>
  Large text
</p>
`;

exports[`Typography - Snapshots > matches snapshot for medium size 1`] = `
<p
  class="text-xl font-semibold text-foreground"
  data-testid="typography"
>
  Medium text
</p>
`;

exports[`Typography - Snapshots > matches snapshot for small size 1`] = `
<p
  class="text-sm font-semibold text-foreground"
  data-testid="typography"
>
  Small text
</p>
`;

exports[`Typography - Snapshots > matches snapshot for span element 1`] = `
<span
  class="text-xl font-semibold text-foreground"
  data-testid="typography"
>
  Span text
</span>
`;

exports[`Typography - Snapshots > matches snapshot for strong element 1`] = `
<strong
  class="text-xl font-semibold text-foreground"
  data-testid="typography"
>
  Strong text
</strong>
`;

exports[`Typography - Snapshots > matches snapshot with combined props 1`] = `
<h2
  class="text-2xl font-bold text-foreground"
  data-testid="typography"
>
  Large heading
</h2>
`;

exports[`Typography - Snapshots > matches snapshot with complex children 1`] = `
<p
  class="text-xl font-semibold text-foreground"
  data-testid="typography"
>
  <span>
    Complex
  </span>
   
  <em>
    typography
  </em>
   content
</p>
`;

exports[`Typography - Snapshots > matches snapshot with custom className 1`] = `
<p
  class="text-xl font-semibold text-foreground custom-typography"
  data-testid="typography"
>
  Custom typography
</p>
`;

exports[`Typography - Snapshots > matches snapshot with data-testid prop 1`] = `
<p
  class="text-xl font-semibold text-foreground"
  data-testid="custom-typography"
>
  Typography with test ID
</p>
`;

exports[`Typography - Snapshots > matches snapshot with default props 1`] = `
<p
  class="text-xl font-semibold text-foreground"
  data-testid="typography"
>
  Default text
</p>
`;

exports[`Typography - Snapshots > matches snapshot with id prop 1`] = `
<p
  class="text-xl font-semibold text-foreground"
  data-testid="typography"
  id="typography-id"
>
  Typography with ID
</p>
`;
